### ISSUE TITLE
Better handling of file-close events on Linux

### ISSUE THREAD

**Issue by nathany** [fsnotify/fsnotify#22]
Requested by @spf13 and @extemporalgenome.

> @spf13: on linux at least, all you need to do the proper tracking of vim/any-other-kind-of-editor "done writing" tracking is watch for IN_CLOSE_WRITE and IN_MOVE_TO on the directory
> 
> @extemporalgenome if it's a lazy editor, that just updates in place, IN_CLOSE_WRITE catches it; if it's a "write to temporary and atomically rename" editor, IN_MOVE_TO catches it

The trouble is in doing this in a cross-platform way. And some users may still want IN_MODIFY (eg. to show the progress of a file writing).

**Comment by stumpyfr** [comment]
Any new on that?

**Comment by nathany** [comment]
Nothing yet.

**Comment by edrex** [comment]
I'm guessing @extemporalgenome's original question was "How can an fsnotify user listen for changes to a single file in a way that works with all [the various text editors' save strategies](https://github.com/guard/guard/wiki/Analysis-of-inotify-events-for-different-editors)?".

Many editors are do a little dance around saving which confounds the strategy of listening to the file for write events. For example, `vim`:

```
2015/03/25 23:18:42 event: "./foo.txt": RENAME
2015/03/25 23:18:42 event: "./foo.txt~": CREATE
2015/03/25 23:18:42 event: "./foo.txt": CREATE
2015/03/25 23:18:42 event: "./foo.txt": WRITE
2015/03/25 23:18:42 modified file: ./foo.txt
```

It seems like the most compatible strategy is to listen to the directory and only pay attention to events matching the file name, but it's not clear which event types to listen to. Just `fsevents.Write` is enough for vim. Any recommendation?

**Comment by nathany** [comment]
This is closely related to #17.

**Comment by rjeczalik** [comment]
Hey @edrex, @nathany! If'd want to use `IN_MOVE_{FROM,TO}` events for inotify, then notify supports them out of the box - see `LinuxMove` example for the [Watch](https://godoc.org/github.com/rjeczalik/notify#Watch) method, in particular:

```go
err := notify.Watch("/dir/to/watch", channel, notify.InMovedFrom, notify.InMovedTo);
```

I'm not suggesting you should pick it over fsnotify, I'm suggesting you could at least test your use-case, whether you really need platform specific events for that.

**Comment by edrex** [comment]
@nathany thanks, #17 is just what I was looking for.

It seems like file watching works across atomic saves on OSX (the watch stays with the path, even though the underlying file is replaced). Curious platform difference..

@rjeczalik I'm really just a user troubleshooting an issue with a downstream package (adnanh/webhook)

**Comment by nathany** [comment]
@rjeczalik Thanks for chiming in. I only heard of your notify project this past week. I'm really interested in taking a closer look at it.

@edrex yah, there are too many little platform differences to trip up on. :dizzy_face:

**Comment by arp242** [comment]
This was essentially fixed with #372 a few months back by providing better documentation and examples; I just missed this issue.
